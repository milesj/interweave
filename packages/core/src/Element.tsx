import React from 'react';

export interface ElementProps {
	[prop: string]: unknown;
	className?: string;
	children?: React.ReactNode;
	selfClose?: boolean;
	tagName: string;
}

export function Element({
	className,
	children = null,
	selfClose = false,
	tagName,
	...props
}: ElementProps) {
	const Tag = tagName as 'span';

	return selfClose ? (
		<Tag className={className} {...props} />
	) : (
		<Tag className={className} {...props}>
			{children}
		</Tag>
	);
}
