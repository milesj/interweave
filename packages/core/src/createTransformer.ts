import { CommonInternals, Node, PassthroughProps, TagName, WildTagName } from './types';

export type InferElement<K> = K extends '*'
	? HTMLElement
	: K extends keyof HTMLElementTagNameMap
	? HTMLElementTagNameMap[K]
	: HTMLElement;

export type TransformerFactory<Element, Props extends object> = (
	element: Element,
	props: Props,
	children: Node[],
) => Element | React.ReactElement | null | undefined | void;

export interface TransformerOptions<Props extends object, Config extends object>
	extends CommonInternals<Props, Config> {
	tagName?: TagName;
}

export interface Transformer<Element, Props extends object, Config extends object>
	extends CommonInternals<Props, Config> {
	extend: (
		config?: Partial<Config>,
		factory?: TransformerFactory<Element, Props>,
	) => Transformer<Element, Props, Config>;
	factory: TransformerFactory<Element, Props>;
	tagName: WildTagName;
}

export function createTransformer<
	K extends WildTagName,
	Props extends object = PassthroughProps,
	Config extends object = {},
>(
	tagName: K,
	factory: TransformerFactory<InferElement<K>, Props>,
	options: TransformerOptions<Props, Config> = {},
): Transformer<InferElement<K>, Props, Config> {
	return {
		config: options.config,
		extend(customConfig, customFactory) {
			return createTransformer<K, Props, Config>(tagName, customFactory ?? factory, {
				...options,
				config: {
					...options.config!,
					...customConfig,
				},
			});
		},
		factory,
		onAfterParse: options.onAfterParse,
		onBeforeParse: options.onBeforeParse,
		tagName: options.tagName ?? tagName,
	};
}
