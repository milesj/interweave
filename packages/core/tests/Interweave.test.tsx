import React from 'react';
import ReactDOMServer from 'react-dom/server';
import { polyfill } from 'interweave-ssr';
import { render } from 'rut-dom';
import { ALLOWED_TAG_LIST } from '../src/constants';
import { createTransformer } from '../src/createTransformer';
import { Element } from '../src/Element';
import { Interweave, InterweaveProps } from '../src/Interweave';
import {
	codeBarMatcher,
	linkTransformer,
	mdBoldMatcher,
	mdItalicMatcher,
	MOCK_INVALID_MARKUP,
	MOCK_MARKUP,
} from '../src/test';

describe('Interweave', () => {
	it('doesnt include canvas and iframe in default allow list', () => {
		expect(ALLOWED_TAG_LIST).not.toContain('canvas');
		expect(ALLOWED_TAG_LIST).not.toContain('iframe');
	});

	it('can pass transformers through props', () => {
		const { root } = render<InterweaveProps>(
			<Interweave content={'Foo <a href="foo.com">Bar</a> Baz'} transformers={[linkTransformer]} />,
		);

		expect(root).toMatchSnapshot();
	});

	it('can pass matchers through props', () => {
		const { root } = render<InterweaveProps>(
			<Interweave content="Foo [bar] Bar Baz" matchers={[codeBarMatcher]} />,
		);

		expect(root).toMatchSnapshot();
	});

	it('allows empty `content` to be passed', () => {
		const { root } = render<InterweaveProps>(<Interweave content={null} />);

		expect(root).toMatchSnapshot();
	});

	it('allows empty `content` to be passed when using callbacks', () => {
		const { root } = render<InterweaveProps>(
			<Interweave content={null} onBeforeParse={(value) => value} />,
		);

		expect(root).toMatchSnapshot();
	});

	it('renders using a custom container element', () => {
		const { root } = render<InterweaveProps>(
			<Interweave content="<li>Foo</li><li>Bar</li><li>Baz</li>" tagName="ul" />,
		);

		expect(root).toMatchSnapshot();
	});

	describe('parseMarkup()', () => {
		it('errors if onBeforeParse doesnt return a string', () => {
			expect(() => {
				// @ts-expect-error Invalid type
				render<InterweaveProps>(<Interweave content="Foo" onBeforeParse={() => 123} />);
			}).toThrowErrorMatchingSnapshot();
		});

		it('can modify the markup using onBeforeParse', () => {
			const { root } = render<InterweaveProps>(
				<Interweave
					content={'Foo <b>Bar</b> Baz'}
					onBeforeParse={(content) => content.replace(/b>/g, 'i>')}
				/>,
			);

			expect(root).toMatchSnapshot();
		});

		it('can modify the tree using onAfterParse', () => {
			const { root } = render<InterweaveProps>(
				<Interweave
					content={'Foo <b>Bar</b> Baz'}
					onAfterParse={(content) => (
						<>
							{content}
							<Element key="1" tagName="u">
								Qux
							</Element>
						</>
					)}
				/>,
			);

			expect(root).toMatchSnapshot();
		});
	});

	describe('parsing and rendering', () => {
		it('handles void elements correctly', () => {
			const { root } = render<InterweaveProps>(
				<Interweave
					content={
						'This has line breaks.<br>Horizontal rule.<hr />An image.<img src="http://domain.com/image.jpg" />'
					}
					tagName="div"
				/>,
			);

			expect(root).toMatchSnapshot();
		});
	});

	describe('line breaks', () => {
		it('converts line breaks', () => {
			const { root } = render<InterweaveProps>(<Interweave content={'Foo\nBar'} />);

			expect(root).toMatchSnapshot();
		});

		it('converts line breaks if `noHtmlExceptInternals` is true', () => {
			const { root } = render<InterweaveProps>(
				<Interweave noHtmlExceptInternals content={'Foo\nBar'} />,
			);

			expect(root).toMatchSnapshot();
		});

		it('doesnt convert line breaks if `noHtml` is true', () => {
			const { root } = render<InterweaveProps>(<Interweave noHtml content={'Foo\nBar'} />);

			expect(root).toMatchSnapshot();
		});

		it('doesnt convert line breaks if `disableLineBreaks` is true', () => {
			const { root } = render<InterweaveProps>(
				<Interweave disableLineBreaks content={'Foo\nBar'} />,
			);

			expect(root).toMatchSnapshot();
		});

		it('doesnt convert line breaks if it contains HTML', () => {
			const { root } = render<InterweaveProps>(<Interweave content={'Foo\n<br/>Bar'} />);

			expect(root).toMatchSnapshot();
		});
	});

	describe('allow list', () => {
		it('filters invalid tags and attributes', () => {
			const { root } = render<InterweaveProps>(<Interweave content={MOCK_INVALID_MARKUP} />);

			expect(root).toMatchSnapshot();
		});

		it('doesnt filter invalid tags and attributes when disabled', () => {
			const { root } = render<InterweaveProps>(
				<Interweave allowAttributes allowElements content={MOCK_INVALID_MARKUP} />,
			);

			expect(root).toMatchSnapshot();
		});
	});

	describe('block list', () => {
		it('filters blocked tags and attributes', () => {
			const { root } = render<InterweaveProps>(
				<Interweave block={['aside', 'a']} content={MOCK_MARKUP} />,
			);

			expect(root).toMatchSnapshot();
		});
	});

	describe('server side rendering', () => {
		let implSpy: jest.SpyInstance;

		beforeEach(() => {
			polyfill();

			implSpy = jest.spyOn(document.implementation, 'createHTMLDocument');
		});

		afterEach(() => {
			global.INTERWEAVE_SSR_POLYFILL = undefined;

			implSpy.mockRestore();
		});

		it('renders basic HTML', () => {
			const actual = ReactDOMServer.renderToStaticMarkup(
				<Interweave content="This is <b>bold</b>." />,
			);

			expect(actual).toBe('This is <b>bold</b>.');
			expect(implSpy).not.toHaveBeenCalled();
		});

		it('strips HTML', () => {
			const actual = ReactDOMServer.renderToStaticMarkup(
				<Interweave noHtml content="This is <b>bold</b>." />,
			);

			expect(actual).toBe('This is bold.');
			expect(implSpy).not.toHaveBeenCalled();
		});

		it('supports styles', () => {
			const actual = ReactDOMServer.renderToStaticMarkup(
				<Interweave content="This is <b style='font-weight: bold'>bold</b>." />,
			);

			expect(actual).toBe('This is <b style="font-weight:bold">bold</b>.');
			expect(implSpy).not.toHaveBeenCalled();
		});

		it('supports transformers', () => {
			const actual = ReactDOMServer.renderToStaticMarkup(
				<Interweave
					content={'Foo <a href="foo.com">Bar</a> Baz'}
					transformers={[linkTransformer]}
				/>,
			);

			expect(actual).toBe('Foo <a href="bar.net" target="_blank">Bar</a> Baz');
			expect(implSpy).not.toHaveBeenCalled();
		});

		it('supports matchers', () => {
			const actual = ReactDOMServer.renderToStaticMarkup(
				<Interweave content="Foo [bar] Bar Baz" matchers={[codeBarMatcher]} />,
			);

			expect(actual).toBe('Foo <span customprop="bar">BAR</span> Bar Baz');
			expect(implSpy).not.toHaveBeenCalled();
		});
	});

	describe('transform prop', () => {
		it('skips the element', () => {
			const transformer = createTransformer('img', () => null);

			const { root } = render<InterweaveProps>(
				<Interweave content={'Foo <img/> Bar'} transformers={[transformer]} />,
			);

			expect(root).toMatchSnapshot();
		});

		it('replaces the element', () => {
			function Dummy() {
				return <div />;
			}

			const transformer = createTransformer('img', () => <Dummy />);

			const { root } = render<InterweaveProps>(
				<Interweave content={'Foo <img/> Bar'} transformers={[transformer]} />,
			);

			expect(root).toMatchSnapshot();
		});

		it('allows blocked', () => {
			function Dummy() {
				return <iframe title="foo" />;
			}

			const transformer = createTransformer('iframe', () => <Dummy />);

			const { root } = render<InterweaveProps>(
				<Interweave content={'Foo <iframe></iframe> Bar'} transformers={[transformer]} />,
			);

			expect(root).toMatchSnapshot();
		});

		it.skip('skips transforming tags outside the allowList when transformOnlyAllowList is true', () => {
			const transformer = createTransformer('*', (element) =>
				element.nodeName === 'A' ? <a href="http://example.com">hi</a> : undefined,
			);

			const content = `
        Hello, https://test.example.com <a href="hi">yo</a> test@example.com test@example.com <br>
        at 8:58 AM Someone <someone@somedomain.dev> wrote:<br>Hi, <br>:  Czech, English, Français,
        Português, Español<br><br>Someone<br><br>On Jun 30, 2021, at 9:40 AM, Someone <someone@somedomain.dev>
        wrote:<br><br><br>Hello Name,<br><br>, . <br>, . <br><br><br>SomeDomain.dev<br>P: 999-999-9999<br>
        Português, Español<br><br><br>On Tue, Jun 29, 2021 at 2:05 PM Someone <someone@somedomain.dev>
        <someone@someotherdomain.dev> wrote:<br>Proin consequat ut metus sit amet cursus.<br>
        <br><br>Someone<br><br>On Jun 29, 2021, at 12:15 PM, Someone <someone@somedomain.dev> wrote:
        <br><br><br>No mam,<br><br>, . <br><br><br>SomeDomain.dev<br>P: 999-999-9999<br>Please, let us
        <br><br><br>On Tue, Jun 29, 2021 at 9:54 AM Someone Else <someone@somethirddomain.dev> wrote:<br>,
        . <br><br>Someone Else<br><br><br><br>On Jun 29, 2021, at 9:41 AM, Someone <someone@somedomain.dev>
        <someone@someotherdomain.dev> wrote:<br>Good Morning, <br>, . <br><br>Thank you, <br><br>Someone
        Else<br><br>On Jun 27, 2021, at 3:56 PM, Someone <someone@somedomain.dev> wrote:<br><br><br>, .
        <br><br><br>SomeDomain.dev<br>P: 999-999-9999<br>, Português, Español<br><br><br>On Sun, Jun 27,
        2021 at 12:09 PM Someone <someone@someotherdomain.dev> wrote:<br><br>, . <br><br><image0.jpeg>
        <br><image1.jpeg><br><image3.jpeg><br><br><br>Someone Else<br><br><br><br>On Jun 27, 2021, at
        11:28 AM, Someone <someone@somedomain.dev> wrote:<br><br><br>Hello,<br><br>, . <br><br><br>
        SomeDomain.dev<br>P: 999-999-9999<br>, Português, Español<br><br><br>On Sat, Jun 26, 2021 at
        3:25 PM Someone <someone@somedomain.dev> wrote:<br>, . <br><br><br>SomeDomain.dev<br>P:
        999-999-9999<br>, Português, Español<br><br><br>On Sat, Jun 26, 2021 at 3:21 PM Someone
        <someone@somedomain.dev> wrote:<br>Hello Name,<br><br>, . <br><br>, . <br><br>, . <br><br>, .
        <br><br><br>SomeDomain.dev<br>P: 999-999-9999<br>, Português, Español<br><br><br>On Sat, Jun 26,
        2021 at 1:51 AM Someone <someone@someotherdomain.dev> wrote:<br>Hello,<br>, . <br><br>Someone
        Else <br><br>(999)999-9999<br><br>On Jun 25, 2021, at 3:56 PM, Someone <someone@somedomain.dev>
        wrote:<br><br><br>Please contact the maintenance supervisor regarding this at 19999999999 <br>
        <br>On Fri, Jun 25, 2021, 10:26 AM Someone <someone@someotherdomain.dev> wrote:<br>rttitor semper.
        <br><br><br><br><br>On Jun 24, 2021, at 10:16 AM, Someone <someone@somedomain.dev> wrote:
        <br><br><br>Hello,<br><br>, . .<br><br><br>SomeDomain.dev<br>P: 999-999-9999<br>, Português,
        Español<br><br><br>
      `;

			const { root } = render<InterweaveProps>(
				<Interweave transformOnlyAllowList content={content} transformers={[transformer]} />,
			);

			expect(root).toMatchSnapshot();
		});
	});

	describe('interleaving', () => {
		it('renders them separately', () => {
			const { root } = render<InterweaveProps>(
				<Interweave
					content="This should be **bold** and this _italic_."
					matchers={[mdBoldMatcher, mdItalicMatcher]}
				/>,
			);

			expect(root).toMatchSnapshot();
		});

		it('renders italic in bold', () => {
			const { root } = render<InterweaveProps>(
				<Interweave
					content="This should be **_italic and bold_**."
					matchers={[mdBoldMatcher, mdItalicMatcher]}
				/>,
			);

			expect(root).toMatchSnapshot();
		});

		it('renders bold in italic', () => {
			const { root } = render<InterweaveProps>(
				<Interweave
					content="This should be _**bold and italic**_."
					matchers={[mdBoldMatcher, mdItalicMatcher]}
				/>,
			);

			expect(root).toMatchSnapshot();
		});
	});
});
