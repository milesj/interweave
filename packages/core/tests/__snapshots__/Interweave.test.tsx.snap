// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interweave allow list doesnt filter invalid tags and attributes when disabled 1`] = `
<Interweave
  allowAttributes
  allowElements
  content="<div bgcolor="black">
    <font color="red">Outdated font.</font>
    <script type="text/javascript"></script>
    <p align="center">More text <strike>with outdated stuff</strike>.</p>
  </div>"
>
  <Element bgcolor="black" tagName="div">
    <div bgcolor="black">
      
  
      <Element color="red" tagName="font">
        <font color="red">
          Outdated font.
        </font>
      </Element>
      
  
      
  
      <Element align="center" tagName="p">
        <p align="center">
          More text 
          <Element tagName="strike">
            <strike>with outdated stuff</strike>
          </Element>
          .
        </p>
      </Element>
      

    </div>
  </Element>
</Interweave>
`;

exports[`Interweave allow list filters invalid tags and attributes 1`] = `
<Interweave
  content="<div bgcolor="black">
    <font color="red">Outdated font.</font>
    <script type="text/javascript"></script>
    <p align="center">More text <strike>with outdated stuff</strike>.</p>
  </div>"
>
  <Element tagName="div">
    <div>
      
  
      Outdated font.
      
  
      
  
      <Element tagName="p">
        <p>More text with outdated stuff.</p>
      </Element>
      

    </div>
  </Element>
</Interweave>
`;

exports[`Interweave allows empty \`content\` to be passed 1`] = `<Interweave />`;

exports[`Interweave allows empty \`content\` to be passed when using callbacks 1`] = `<Interweave onBeforeParse={onBeforeParse()} />`;

exports[`Interweave block list filters blocked tags and attributes 1`] = `
<Interweave
  block={[ "aside", "a" ]}
  content="<main role="main">
    Main content
    <div>
      <a href="#">Link</a>
      <span class="foo">String</span>
    </div>
  </main>
  <aside id="sidebar">
    Sidebar content
  </aside>"
>
  <Element role="main" tagName="main">
    <main role="main">
      
  Main content
  
      <Element tagName="div">
        <div>
          
    
          Link
          
    
          <Element className="foo" tagName="span">
            <span className="foo">
              String
            </span>
          </Element>
          
  
        </div>
      </Element>
      

    </main>
  </Element>
  

  
  Sidebar content

</Interweave>
`;

exports[`Interweave can pass matchers through props 1`] = `
<Interweave content="Foo [bar] Bar Baz" matchers={[ <span /> ]}>
  Foo 
  <Element customprop="bar" tagName="span">
    <span customprop="bar">
      BAR
    </span>
  </Element>
   Bar Baz
</Interweave>
`;

exports[`Interweave can pass transformers through props 1`] = `
<Interweave
  content="Foo <a href="foo.com">Bar</a> Baz"
  transformers={[ <a />, <* /> ]}
>
  Foo 
  <Element href="bar.net" tagName="a" target="_blank">
    <a href="bar.net" target="_blank">
      Bar
    </a>
  </Element>
   Baz
</Interweave>
`;

exports[`Interweave interleaving renders bold in italic 1`] = `
<Interweave
  content="This should be _**bold and italic**_."
  matchers={[ <b />, <i /> ]}
>
  This should be 
  <i>
    <b>bold and italic</b>
  </i>
  .
</Interweave>
`;

exports[`Interweave interleaving renders italic in bold 1`] = `
<Interweave
  content="This should be **_italic and bold_**."
  matchers={[ <b />, <i /> ]}
>
  This should be 
  <b>
    <i>italic and bold</i>
  </b>
  .
</Interweave>
`;

exports[`Interweave interleaving renders them separately 1`] = `
<Interweave
  content="This should be **bold** and this _italic_."
  matchers={[ <b />, <i /> ]}
>
  This should be 
  <b>bold</b>
   and this 
  <i>italic</i>
  .
</Interweave>
`;

exports[`Interweave line breaks converts line breaks 1`] = `
<Interweave
  content="Foo
  Bar"
>
  Foo
  <Element selfClose tagName="br">
    <br />
  </Element>
  Bar
</Interweave>
`;

exports[`Interweave line breaks converts line breaks if \`noHtmlExceptInternals\` is true 1`] = `
<Interweave
  noHtmlExceptInternals
  content="Foo
  Bar"
>
  Foo
  <Element selfClose tagName="br">
    <br />
  </Element>
  Bar
</Interweave>
`;

exports[`Interweave line breaks doesnt convert line breaks if \`disableLineBreaks\` is true 1`] = `
<Interweave
  disableLineBreaks
  content="Foo
  Bar"
>
  Foo
Bar
</Interweave>
`;

exports[`Interweave line breaks doesnt convert line breaks if \`noHtml\` is true 1`] = `
<Interweave
  noHtml
  content="Foo
  Bar"
>
  Foo
Bar
</Interweave>
`;

exports[`Interweave line breaks doesnt convert line breaks if it contains HTML 1`] = `
<Interweave
  content="Foo
  <br/>Bar"
>
  Foo

  <Element selfClose tagName="br">
    <br />
  </Element>
  Bar
</Interweave>
`;

exports[`Interweave parseMarkup() can modify the markup using onBeforeParse 1`] = `
<Interweave content="Foo <b>Bar</b> Baz" onBeforeParse={onBeforeParse()}>
  Foo 
  <Element tagName="i">
    <i>Bar</i>
  </Element>
   Baz
</Interweave>
`;

exports[`Interweave parseMarkup() can modify the tree using onAfterParse 1`] = `
<Interweave content="Foo <b>Bar</b> Baz" onAfterParse={onAfterParse()}>
  Foo 
  <Element tagName="b">
    <b>Bar</b>
  </Element>
   Baz
  <Element tagName="u">
    <u>Qux</u>
  </Element>
</Interweave>
`;

exports[`Interweave parseMarkup() errors if onBeforeParse doesnt return a string 1`] = `"Interweave \`onBeforeParse\` must return a valid HTML string."`;

exports[`Interweave parsing and rendering handles void elements correctly 1`] = `
<Interweave
  content="This has line breaks.<br>Horizontal rule.<hr />An image.<img src="http://domain.com/image.jpg" />"
  tagName="div"
>
  This has line breaks.
  <Element selfClose tagName="br">
    <br />
  </Element>
  Horizontal rule.
  <Element selfClose tagName="hr">
    <hr />
  </Element>
  An image.
  <Element selfClose src="http://domain.com/image.jpg" tagName="img">
    <img src="http://domain.com/image.jpg" />
  </Element>
</Interweave>
`;

exports[`Interweave renders using a custom container element 1`] = `
<Interweave content="<li>Foo</li><li>Bar</li><li>Baz</li>" tagName="ul">
  <Element tagName="li">
    <li>Foo</li>
  </Element>
  <Element tagName="li">
    <li>Bar</li>
  </Element>
  <Element tagName="li">
    <li>Baz</li>
  </Element>
</Interweave>
`;

exports[`Interweave transform prop allows blocked 1`] = `
<Interweave
  content="Foo <iframe></iframe> Bar"
  transformers={[ <iframe />, <* /> ]}
>
  Foo 
  <Dummy>
    <iframe title="foo" />
  </Dummy>
   Bar
</Interweave>
`;

exports[`Interweave transform prop replaces the element 1`] = `
<Interweave content="Foo <img/> Bar" transformers={[ <img />, <* /> ]}>
  Foo 
  <Dummy>
    <div />
  </Dummy>
   Bar
</Interweave>
`;

exports[`Interweave transform prop skips the element 1`] = `
<Interweave content="Foo <img/> Bar" transformers={[ <img />, <* /> ]}>
  Foo 
   Bar
</Interweave>
`;

exports[`Interweave transform prop skips transforming tags outside the allowList when transformOnlyAllowList is true 1`] = `
<Element tagName="span">
  <span>

        Hello, https://test.example.com
    <a href="http://example.com">
      hi
    </a>
     test@example.com test@example.com
    <Element selfClose tagName="br">
      <br />
    </Element>

        at 8:58 AM Someone
     wrote:
    <Element selfClose tagName="br">
      <br />
    </Element>
    Hi,
    <Element selfClose tagName="br">
      <br />
    </Element>
    :  Czech, English, Français,
        Português, Español
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    Someone
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    On Jun 30, 2021, at 9:40 AM, Someone

        wrote:
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    Hello Name,
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    , .
    <Element selfClose tagName="br">
      <br />
    </Element>
    , .
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    SomeDomain.dev
    <Element selfClose tagName="br">
      <br />
    </Element>
    P: 999-999-9999
    <Element selfClose tagName="br">
      <br />
    </Element>

        Português, Español
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    On Tue, Jun 29, 2021 at 2:05 PM Someone


     wrote:
    <Element selfClose tagName="br">
      <br />
    </Element>
    Proin consequat ut metus sit amet cursus.
    <Element selfClose tagName="br">
      <br />
    </Element>


    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    Someone
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    On Jun 29, 2021, at 12:15 PM, Someone
     wrote:

    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    No mam,
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    , .
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    SomeDomain.dev
    <Element selfClose tagName="br">
      <br />
    </Element>
    P: 999-999-9999
    <Element selfClose tagName="br">
      <br />
    </Element>
    Please, let us

    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    On Tue, Jun 29, 2021 at 9:54 AM Someone Else
     wrote:
    <Element selfClose tagName="br">
      <br />
    </Element>
    ,
        .
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    Someone Else
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    On Jun 29, 2021, at 9:41 AM, Someone


     wrote:
    <Element selfClose tagName="br">
      <br />
    </Element>
    Good Morning,
    <Element selfClose tagName="br">
      <br />
    </Element>
    , .
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    Thank you,
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    Someone
        Else
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    On Jun 27, 2021, at 3:56 PM, Someone
     wrote:
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    , .

    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    SomeDomain.dev
    <Element selfClose tagName="br">
      <br />
    </Element>
    P: 999-999-9999
    <Element selfClose tagName="br">
      <br />
    </Element>
    , Português, Español
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    On Sun, Jun 27,
        2021 at 12:09 PM Someone
     wrote:
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    , .
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>


    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    Someone Else
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    On Jun 27, 2021, at
        11:28 AM, Someone
     wrote:
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    Hello,
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    , .
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>

        SomeDomain.dev
    <Element selfClose tagName="br">
      <br />
    </Element>
    P: 999-999-9999
    <Element selfClose tagName="br">
      <br />
    </Element>
    , Português, Español
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    On Sat, Jun 26, 2021 at
        3:25 PM Someone
     wrote:
    <Element selfClose tagName="br">
      <br />
    </Element>
    , .
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    SomeDomain.dev
    <Element selfClose tagName="br">
      <br />
    </Element>
    P:
        999-999-9999
    <Element selfClose tagName="br">
      <br />
    </Element>
    , Português, Español
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    On Sat, Jun 26, 2021 at 3:21 PM Someone

     wrote:
    <Element selfClose tagName="br">
      <br />
    </Element>
    Hello Name,
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    , .
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    , .
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    , .
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    , .

    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    SomeDomain.dev
    <Element selfClose tagName="br">
      <br />
    </Element>
    P: 999-999-9999
    <Element selfClose tagName="br">
      <br />
    </Element>
    , Português, Español
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    On Sat, Jun 26,
        2021 at 1:51 AM Someone
     wrote:
    <Element selfClose tagName="br">
      <br />
    </Element>
    Hello,
    <Element selfClose tagName="br">
      <br />
    </Element>
    , .
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    Someone
        Else
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    (999)999-9999
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    On Jun 25, 2021, at 3:56 PM, Someone

        wrote:
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    Please contact the maintenance supervisor regarding this at 19999999999
    <Element selfClose tagName="br">
      <br />
    </Element>


    <Element selfClose tagName="br">
      <br />
    </Element>
    On Fri, Jun 25, 2021, 10:26 AM Someone
     wrote:
    <Element selfClose tagName="br">
      <br />
    </Element>
    rttitor semper.

    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    On Jun 24, 2021, at 10:16 AM, Someone
     wrote:

    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    Hello,
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    , . .
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    SomeDomain.dev
    <Element selfClose tagName="br">
      <br />
    </Element>
    P: 999-999-9999
    <Element selfClose tagName="br">
      <br />
    </Element>
    , Português,
        Español
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>
    <Element selfClose tagName="br">
      <br />
    </Element>


  </span>
</Element>
`;
